### ISSUE TITLE
test-low-level-api.jl fails due to 6-bit config test

### ISSUE THREAD

**Issue by mgkuhn** [JuliaIO/LibSerialPort.jl#55]
I noticed that `julia test/test-low-level-api.jl /dev/ttyUSB0 115200` outputs string `"\x14%34 -%33!'% 1\n"` instead of the expected `Test message 1`, `Test message 2`, etc. I ran the test while I had arranged a loop-back configuration with my serial-port adapter (e.g. [Adafruit USB to TTL serial cable](https://www.adafruit.com/product/954) with its white/green RX/TX lines simply connected with each other with a piece of wire).

I then got it to work (via trial and error) by uncommenting in `function test_low_level_api` of [test/test-low-level-api.jl](https://github.com/JuliaIO/LibSerialPort.jl/blob/f05a87ba458218378078e6d4c63279750ff135d4/test/test-low-level-api.jl) the call to `test_port_configuration(port)`. That function appears to leaves the `port` handle in an unusable state, despite it claiming in the doc string: `The original port configuration should not be modified by these tests!`.

Something seems very confused here: both `test_change_port_copy_method1()` and `test_change_port_copy_method2()` each have in their second line `sp_close(port)`, while the rest of `test_low_level_api()` afterwards happily tries to continue using the `port` handle that has been closed twice!

How was this meant to work?

**Comment by andrewadare** [comment]
I can't reproduce this problem, but I'm working on master after your PR #56, and I didn't take the time to check out an older revision and troubleshoot.  Since I wrote these tests 4 years ago, the tests have not received adequate attention to track with changes in libserialport, Julia, and this package. I am working on improvements currently and would be grateful for your review when ready.

However, I'm not sure if I understand how the logic is confused: during the tests, the file descriptor for the original port is closed, but that does not delete the underlying `sp_port` or `sp_port_config` structs. At the end of the tests, the port is reopened, so the handle is available for use later.

**Comment by mgkuhn** [comment]
It appears that the two lines
```
sp_set_bits(port2, 6)
```
and
```
sp_set_config_bits(config2, 6)
```
in the functions called by `test_port_configuration()` have a persistent effect, not just within the `test/test-low-level-api.jl` process, but even across invocations of that script. They appear to persistently configure my [USB serial port adapter `/dev/serial/by-id/usb-Silicon_Labs_CP2102_USB_to_UART_Bridge_Controller_0001-if00-port0`](https://www.adafruit.com/product/954) into 6-bit mode, and its stays that way until I explicitly configure it to 8-bit mode (e.g., by opening it with a high-level interface routine like `LibSerialPort.open_serial_port` that does that). That's on Ubuntu 18.04 ix86_64.

**Comment by andrewadare** [comment]
To troubleshoot this, in #61 I added the explicit check
```julia
@test get_port_settings(port) == original_port_settings
```
at the end of both `test_change_port_copy_method1()` and `test_change_port_copy_method2()`. These tests pass for me (as do all subsequent tests), so I'm at a bit of a loss. I would be interested to know whether this is the case with your setup.

**Comment by mgkuhn** [comment]
Your new check fails for me the first time, but not the second time, see log of two consecutive invocations below. I think I now understand what is going on.

If you are only using an Arduino, you do not actually have a real serial-port adapter. Nothing in your system actually produces serialized bytes in the requested serial-bus frame format on a single wire. Your Arduino probably just pretends to be a serial port UART, but actually just copies bytes received via USB transport mode into microcontroller memory. Therefore, your Arduino probably completely ignores any of your attempts to set a UART to output a certain number of data bits, parity bits, number of stop bits, possibly even the data rate. I am instead using a real serial-port adapter, where I can see bytes with start and stop bits flying past on a wire (with a logic analyzer).

I suspect the problem may be a missunderstanding of how libserialports controls the data format registers in the UART. I assume that when you call `sp_set_bits(port2, 6)`, that call actually instantly configures the UART hardware into a mode where it now outputs only 6 data bits per word. After that, your test never does anything to restore that UART configuration register back to 8 data bits per word. You believe that you are modifying a copy of the UART configuration, but since there is actually only one UART hardware register underneath, both the `port` handle and its `port2` copy send the 6-bit configuration change to the same hardware, and after that never revert it.

I therefore suspect your Arduino configuration is unsuitable as a test device, as it does not actually involve anything resembling a real UART.

Note how in the second test below, my CP210x UART Bridge actually has remembered the 6-bit setting across two Julia processes!

```
$ ls -l /dev/serial/by-id/usb-Silicon_Labs_CP2102_USB_to_UART_Bridge_Controller_0001-if00-port0
lrwxrwxrwx 1 root root 13 Apr 22 13:15 /dev/serial/by-id/usb-Silicon_Labs_CP2102_USB_to_UART_Bridge_Controller_0001-if00-port0 -> ../../ttyUSB0
$ lsusb -d 10c4:ea60
Bus 002 Device 006: ID 10c4:ea60 Cygnal Integrated Products, Inc. CP210x UART Bridge / myAVR mySmartUSB light
$ julia test/test-low-level-api.jl /dev/ttyUSB0 115200
0
1
1
0.1.1
1
0
1
1:0:1
/dev/ttyUSB1
	Description:	Digilent USB Device - 210321A6A8A3
	Transport type:	SP_TRANSPORT_USB
/dev/ttyS0
	Description:	ttyS0
	Transport type:	SP_TRANSPORT_NATIVE
/dev/ttyUSB0
	Description:	CP2102 USB to UART Bridge Controller - 0001
	Transport type:	SP_TRANSPORT_USB

[TEST1] INITIAL Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_OUT
	stopbits	1
	RTS	SP_RTS_ON
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_INVALID
	bits	8
	DTR	SP_DTR_ON
	CTS	SP_CTS_IGNORE
[TEST1] changing port configuration settings.
[TEST1] UPDATED Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_OFF
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_OFF
	CTS	SP_CTS_IGNORE
[TEST1] closing and deleting copied port
[TEST1] reopening original port
[TEST1] ORIGINAL Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_ON
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_ON
	CTS	SP_CTS_IGNORE
Test port configuration: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:73
  Expression: get_port_settings(port) == original_port_settings
   Evaluated: Dict{String,Any}("XonXoff" => SP_XONXOFF_INOUT,"stopbits" => 2,"RTS" => SP_RTS_ON,"baudrate" => 115200,"DSR" => SP_DSR_IGNORE,"parity" => SP_PARITY_EVEN,"bits" => 6,"DTR" => SP_DTR_ON,"CTS" => SP_CTS_IGNORE) == Dict{String,Any}("XonXoff" => SP_XONXOFF_OUT,"stopbits" => 1,"RTS" => SP_RTS_ON,"baudrate" => 115200,"DSR" => SP_DSR_IGNORE,"parity" => SP_PARITY_INVALID,"bits" => 8,"DTR" => SP_DTR_ON,"CTS" => SP_CTS_IGNORE)
Stacktrace:
 [1] test_change_port_copy_method1(::Ptr{LibSerialPort.SPPort}) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:73
 [2] test_port_configuration at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:19 [inlined]
 [3] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:268 [inlined]
 [4] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [5] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:268

[TEST2] INITIAL 	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_ON
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_ON
	CTS	SP_CTS_IGNORE
[TEST2] changing configuration settings.
[TEST2] UPDATED Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_OFF
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_OFF
	CTS	SP_CTS_IGNORE
[TEST2] closing and deleting copied port
[TEST2] reopening original port
[TEST2] ORIGINAL Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_ON
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_ON
	CTS	SP_CTS_IGNORE
Test Summary:           | Pass  Fail  Total
Test port configuration |    1     1      2


$ julia test/test-low-level-api.jl /dev/ttyUSB0 115200
0
1
1
0.1.1
1
0
1
1:0:1
/dev/ttyUSB1
	Description:	Digilent USB Device - 210321A6A8A3
	Transport type:	SP_TRANSPORT_USB
/dev/ttyS0
	Description:	ttyS0
	Transport type:	SP_TRANSPORT_NATIVE
/dev/ttyUSB0
	Description:	CP2102 USB to UART Bridge Controller - 0001
	Transport type:	SP_TRANSPORT_USB

[TEST1] INITIAL Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_ON
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_ON
	CTS	SP_CTS_IGNORE
[TEST1] changing port configuration settings.
[TEST1] UPDATED Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_OFF
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_OFF
	CTS	SP_CTS_IGNORE
[TEST1] closing and deleting copied port
[TEST1] reopening original port
[TEST1] ORIGINAL Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_ON
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_ON
	CTS	SP_CTS_IGNORE

[TEST2] INITIAL 	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_ON
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_ON
	CTS	SP_CTS_IGNORE
[TEST2] changing configuration settings.
[TEST2] UPDATED Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_OFF
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_OFF
	CTS	SP_CTS_IGNORE
[TEST2] closing and deleting copied port
[TEST2] reopening original port
[TEST2] ORIGINAL Configuration for /dev/ttyUSB0:
	XonXoff	SP_XONXOFF_INOUT
	stopbits	2
	RTS	SP_RTS_ON
	baudrate	115200
	DSR	SP_DSR_IGNORE
	parity	SP_PARITY_EVEN
	bits	6
	DTR	SP_DTR_ON
	CTS	SP_CTS_IGNORE
Test Summary:           | Pass  Total
Test port configuration |    2      2

Testing serial loopback with blocking write/read functions...
Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 1", "\x14%34 -%33!'% 1\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(1) 15, 15
%34 -%33!'% 1

Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 2", "\x14%34 -%33!'% 2\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(2) 15, 15
%34 -%33!'% 2

Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 3", "\x14%34 -%33!'% 3\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(3) 15, 15
%34 -%33!'% 3

Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 4", "\x14%34 -%33!'% 4\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(4) 15, 15
%34 -%33!'% 4

Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 5", "\x14%34 -%33!'% 5\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(5) 15, 15
%34 -%33!'% 5

Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 6", "\x14%34 -%33!'% 6\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(6) 15, 15
%34 -%33!'% 6

Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 7", "\x14%34 -%33!'% 7\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(7) 15, 15
%34 -%33!'% 7

Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 8", "\x14%34 -%33!'% 8\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(8) 15, 15
%34 -%33!'% 8

Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 9", "\x14%34 -%33!'% 9\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(9) 15, 15
%34 -%33!'% 9

Blocking read/write: Test Failed at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
  Expression: occursin("Test message $(i)", response)
   Evaluated: occursin("Test message 10", "\x14%34 -%33!'% 10\n")
Stacktrace:
 [1] (::var"#loopback_with_reallocation#3"{Ptr{LibSerialPort.SPPort},Int64,Int64})() at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:170
 [2] macro expansion at ./util.jl:175 [inlined]
 [3] test_blocking_serial_loopback(::Ptr{LibSerialPort.SPPort}, ::Int64, ::Int64) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:207
 [4] macro expansion at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276 [inlined]
 [5] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [6] test_low_level_api(::String, ::Vararg{String,N} where N) at /home/mgk25/.julia/dev/LibSerialPort/test/test-low-level-api.jl:276
(10) 16, 16
%34 -%33!'% 10

  3.115939 seconds (3.97 M allocations: 189.096 MiB, 3.65% gc time)
(1) 15, 15
%34 -%33!'% 1

(2) 15, 15
%34 -%33!'% 2

(3) 15, 15
%34 -%33!'% 3

(4) 15, 15
%34 -%33!'% 4

(5) 15, 15
%34 -%33!'% 5

(6) 15, 15
%34 -%33!'% 6

(7) 15, 15
%34 -%33!'% 7

(8) 15, 15
%34 -%33!'% 8

(9) 15, 15
%34 -%33!'% 9

(10) 16, 16
%34 -%33!'% 10

  1.012652 seconds (310 allocations: 26.125 KiB)
Test Summary:       | Pass  Fail  Total
Blocking read/write |   20    10     30
```

**Comment by mgkuhn** [comment]
Some low-cost serial-port adapters and accessories that might be useful for loop-back testing with a real UART:
* https://www.adafruit.com/product/954 – that's the one I am using currently. Advantage: TTL output level is compatible with logic analysers, disadvantage: it lacks modem control pins and offers only RX + TX wires (plus 5V power and ground)
* https://www.adafruit.com/product/4364 – that one has in addition the RTS/CTS flow control wires, but still lacks the other modem-control signals
* https://www.amazon.com/Sabrent-Converter-Prolific-Chipset-CB-DB9P/dp/B00IDSM6BW – looks like a proper RS-232 adapter with all DE-9 modem control signals
* real RS-232 adapters are best used with an LED indicator such as https://www.amazon.com/TENMA-72-9265-CABLE-TESTER-SIGNAL/dp/B00HPLZD5E
* https://www.amazon.com/Oiyagai-Female-Terminal-Connector-Signal/dp/B074Z55GPN – female DE-9 connector housing to make your own loop-back test adapter (feeding all outputs back onto the corresponding input, as in http://www.bb-elec.com/Learning-Center/All-White-Papers/Serial/RS-232-Loopback-Connections-DB9-DB25.aspx )
* https://www.amazon.co.uk/Sienoc-Express-controller-Adapter-Converter/dp/B00GT65XME – PCIe RS-232 adapter (i.e., no USB involved, proper PC COM port, but your motherboard may already have one)

**Comment by mgkuhn** [comment]
I've seen [one data sheet for the CH340 UART](https://cdn.sparkfun.com/datasheets/Dev/Arduino/Other/CH340DS1.PDF) used on some Arduino boards that claims only support for 8 or 9 data bits, not 6.

**Comment by andrewadare** [comment]
> You believe that you are modifying a copy of the UART configuration, but since there is actually only one UART hardware register underneath, both the port handle and its port2 copy send the 6-bit configuration change to the same hardware, and after that never revert it.

This explanation for what's going on makes sense. I had a fundamental misunderstanding, and logic in the tests appears to be flawed. 

I also see your point that the Arduino Uno has a separate microcontroller on board (Atmega16U2) programmed to do the USB-to-serial conversion in its CPU, which does not involve a real UART.

The thing is, I also tried this with a usb-to-serial adapter with TX and RX connected with a jumper wire. Specifically, it's a Sparkfun [FTDI basic](https://www.sparkfun.com/products/9716) board with an FT232 chip containing "real" UART hardware that interfaces with the AVR micro using TTL serial protocol (at least that's my best understanding). This seems quite similar to one of the Adafruit products you listed above, and I'm not sure why this setup didn't have the same problem as your Si Labs model.

**Comment by andrewadare** [comment]
From the feature list in the FT232 datasheet:
> UART interface support for 7 or 8 data bits, 1
or 2 stop bits and odd / even / mark / space /
no parity

So the 6 data bit configuration looks unsupported. But I tried `sp_set_config_bits(config2, 7)`, and the tests still pass.

**Comment by mgkuhn** [comment]
I have just confirmed that my Silicon Labs CP2102 UART does indeed support 6-bit transmission and reception: if I send out
```
msg = collect(UInt8(0):UInt8(255))
```
I get back
```
msg .& 0x3f
```
i.e. the transmitter only sends the six least significant bits of each byte provided.

Test code for ndatabits=6:
```
using LibSerialPort
msg = collect(UInt8(0):UInt8(255))
dev = "/dev/ttyUSB0"
s = LibSerialPort.open(dev, 115200, ndatabits=6)    # <=== 6-bit mode
set_flow_control(s, xonxoff = SP_XONXOFF_DISABLED)  # otherwise we lose 0x11 0x13
write(s, msg)
sp_drain(s)
rcv = nonblocking_read(s)
msg .&= 0x3f        # <=== expected bytes in 6-bit mode
if rcv == msg
    println("All 256 bytes received")
else
    @warn("Received data does not match transmitted data:")
    println("Bytes received: ", rcv)
    println("Bytes missing:  ", setdiff(msg, rcv));
    println("Bytes added:    ", setdiff(rcv, msg));
end
```

Perhaps you want to include such a binary round-trip test as well (at least for ndatabits=8)?

**Comment by mgkuhn** [comment]
Note that your test messages so far only contain 7-bit ASCII data, therefore you only rely on 7-bit transparency of the connection and wouldn't notice if the UART is still in 7-bit mode. Add an 8-bit 256-byte round-trip test, as above, and you may detect the problem then.

**Comment by andrewadare** [comment]
This is a nice test. With it, I confirmed that your test fails as written above for 6 bits (since my device doesn't support that), but passes with `s = LibSerialPort.open(dev, 115200, ndatabits=7)` and `msg .&= 0x7f`.

I also wrote bytes higher than 0x7f in the test messages as you suggested and finally got `test-low-level-api.jl` to fail, reproducing this issue.

So my testing hardware happened not to catch the problems with the test, but now that I see what you've pointed out I will fix it. Thanks again, very sharp troubleshooting.
